kotlinx-coroutines-test/common/src/TestBuilders.kt

```diff
@@ -350,59 +385,115 @@ internal suspend fun <T: AbstractCoroutine<Unit>> CoroutineScope.runTestCoroutin
     coroutine.start(CoroutineStart.UNDISPATCHED, coroutine) {
         testBody()
     }
-    /**
-     * The general procedure here is as follows:
-     * 1. Try running the work that the scheduler knows about, both background and foreground.
-     *
-     * 2. Wait until we run out of foreground work to do. This could mean one of the following:
-     *    * The main coroutine is already completed. This is checked separately; then we leave the procedure.
-     *    * It's switched to another dispatcher that doesn't know about the [TestCoroutineScheduler].
-     *    * Generally, it's waiting for something external (like a network request, or just an arbitrary callback).
-     *    * The test simply hanged.
-     *    * The main coroutine is waiting for some background work.
-     *
-     * 3. We await progress from things that are not the code under test:
-     *    the background work that the scheduler knows about, the external callbacks,
-     *    the work on dispatchers not linked to the scheduler, etc.
-     *
-     *    When we observe that the code under test can proceed, we go to step 1 again.
-     *    If there is no activity for [dispatchTimeoutMs] milliseconds, we consider the test to have hanged.
-     *
-     *    The background work is not running on a dedicated thread.
-     *    Instead, the test thread itself is used, by spawning a separate coroutine.
-     */
-    var completed = false
-    while (!completed) {
-        scheduler.advanceUntilIdle()
-        if (coroutine.isCompleted) {
-            /* don't even enter `withTimeout`; this allows to use a timeout of zero to check that there are no
+    if (dispatchTimeout != null) {
+        check(testTimeout == null) { "Only one of dispatchTimeout and testTimeout can be specified" }
+        /**
```

Comment:
This is a bit opionated, so feel free to ignore.

I found that extracting this branch's content into  separate `runTestCoroutineLegacy` is much more straightforward -- it immediately becomes clear which code is meaningful and should be future-proof and which is not so. Also, simplifies the body for readers that do not care about legacy behaviour